simulation/src/simulation.rs

```diff
@@ -0,0 +1,362 @@
+                                        panic!("unexpected client error {:?}", error);
+                                    }
+                                }
+                            };
+
+                            *status_codes
+                                .borrow_mut()
+                                .entry(status.as_u16())
+                                .or_insert(0) += 1;
+                        }
+                    }
+                });
+
+                match abort_after {
+                    Some(abort_after) => {
+                        let _ = time::timeout(abort_after, run_requests).await;
+                    }
+                    None => run_requests.await,
+                }
+
+                recorder.lock().record();
+
+                let status_codes = status_codes.into_inner();
+                SimulationReport {
+                    end_time: start.elapsed(),
+                    client_mean: Duration::from_nanos(
+                        metrics::responses_timer(&metrics, SERVICE)
+                            .snapshot()
+                            .mean() as u64,
+                    ),
+                    success_percentage: f64::round(
+                        status_codes.get(&200).copied().unwrap_or(0) as f64 * 1000.
+                            / num_sent.get() as f64,
+                    ) / 10.,
+                    server_cpu: Duration::from_nanos(
+                        metrics::global_server_time_nanos(&metrics).count() as u64,
+                    ),
+                    status_codes,
+                    num_sent: num_sent.get(),
+                    num_received: num_received.get(),
+                    num_global_responses: metrics::global_responses(&metrics).count(),
+                    record: recorder.lock().finish(),
+                }
+            }
+        })
+    }
+}
+
+#[derive(Copy, Clone)]
+pub enum Strategy {
+    ConcurrencyLimiterRoundRobin,
+    ConcurrencyLimiterPinUntilError,
+    UnlimitedRoundRobin,
+}
+
+impl fmt::Display for Strategy {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            Strategy::ConcurrencyLimiterRoundRobin => "CONCURRENCY_LIMITER_ROUND_ROBIN",
+            Strategy::ConcurrencyLimiterPinUntilError => "CONCURRENCY_LIMITER_PIN_UNTIL_ERROR",
+            Strategy::UnlimitedRoundRobin => "UNLIMITED_ROUND_ROBIN",
+        };
+
+        fmt::Display::fmt(s, fmt)
+    }
+}
+
+impl Strategy {
+    fn apply(&self, builder: &mut Builder) {
+        match self {
+            Strategy::ConcurrencyLimiterRoundRobin => {
+                builder.node_selection_strategy(NodeSelectionStrategy::Balanced);
+            }
+            Strategy::ConcurrencyLimiterPinUntilError => {
+                builder.node_selection_strategy(NodeSelectionStrategy::PinUntilError);
+            }
+            Strategy::UnlimitedRoundRobin => {
+                // FIXME disable qos
```

Comment:
What's preventing us from disabling qos here?

Reply:
It's not implemented yet!